python/dgl/data/heterophilous_graphs.py

```diff
@@ -0,0 +1,437 @@
+"""
+Datasets introduced in the 'A Critical Look at the Evaluation of GNNs under Heterophily: Are We
+Really Making Progress? <https://arxiv.org/abs/2302.11640>'__ paper.
+"""
+import os
+
+import numpy as np
+
+from ..convert import graph
+from ..transforms.functional import to_bidirected
+from .dgl_dataset import DGLBuiltinDataset
+from .utils import download
+
+
+class HeterophilousGraphDataset(DGLBuiltinDataset):
+    r"""Datasets introduced in the 'A Critical Look at the Evaluation of GNNs under Heterophily:
+    Are We Really Making Progress? <https://arxiv.org/abs/2302.11640>'__ paper.
+
+    Parameters
+    ----------
+    name : str
+        Name of the dataset. One of 'roman-empire', 'amazon-ratings', 'minesweeper', 'tolokers',
+        'questions'.
+    raw_dir : str
+        Raw file directory to store the processed data.
+    force_reload : bool
+        Whether to re-download the data source.
+    verbose : bool
+        Whether to print progress information.
+    transform : callable
+        A transform that takes in a :class:`~dgl.DGLGraph` object and returns
+        a transformed version. The :class:`~dgl.DGLGraph` object will be
+        transformed before every access.
+    """
+
+    def __init__(self, name, raw_dir=None, force_reload=False, verbose=True, transform=None):
+        name = name.lower().replace('-', '_')
+        url = f'https://github.com/yandex-research/heterophilous-graphs/raw/main/data/{name}.npz'
+        super(HeterophilousGraphDataset, self).__init__(
+            name=name,
+            url=url,
+            raw_dir=raw_dir,
+            force_reload=force_reload,
+            verbose=verbose,
+            transform=transform,
+        )
+
+    def download(self):
+        download(url=self.url, path=os.path.join(self.raw_path, f'{self.name}.npz'))
+
+    def process(self):
+        """Load and process the data."""
+        try:
+            import torch
+        except ImportError:
+            raise ModuleNotFoundError(
+                "This dataset requires PyTorch to be the backend."
+            )
+
+        data = np.load(os.path.join(self.raw_path, f'{self.name}.npz'))
+        src = torch.from_numpy(data['edges'][:, 0])
+        dst = torch.from_numpy(data['edges'][:, 1])
+        features = torch.from_numpy(data['node_features'])
+        labels = torch.from_numpy(data['node_labels'])
+        train_masks = torch.from_numpy(data['train_masks'].T)
+        val_masks = torch.from_numpy(data['val_masks'].T)
+        test_masks = torch.from_numpy(data['test_masks'].T)
+        num_nodes = len(labels)
+        num_classes = len(labels.unique())
+
+        self._num_classes = num_classes
+
+        self._g = to_bidirected(graph((src, dst), num_nodes=num_nodes))
```

Comment:
Are these graphs inherently undirected? If some graphs are inherently directed, then maybe it's better to let users use a transform if needed. Recently, there has been some arising interest in studying directed heterophilous graphs, for example https://arxiv.org/abs/2305.10498

Reply:
All the graphs are undirected. However, in the original files all edges are stored only once (as if in, for example, networkx), thus to_bidirected is required.  There is no point in leaving an option to load the graph without to_bidirected, because than the edge directions will be random.

Reply:
I see. Thanks.